File: 0498-diagonal-traverse/0498-diagonal-traverse.java
class Solution {
    public int[] findDiagonalOrder(int[][] mat) {
        int m=mat.length,n=mat[0].length;
        int []res=new int[m*n];
        boolean up=true;
        int row=0,col=0,i=0;
        while(row<m && col<n){
            if(up){
                while(row>0 && col<n-1){
                    res[i++]=mat[row][col];
                    row--;
                    col++;
                }
                res[i++]=mat[row][col];
                if(col==n-1)row++;
                else col++;
            }
            else{
                while(col>0 && row<m-1){
                    res[i++]=mat[row][col];
                    col--;
                    row++;
                }
                res[i++]=mat[row][col];
                if(row==m-1)col++;
                else row++;
            }
            up=!up;
        }
        return res;
    }
}

File: 0498-diagonal-traverse/0498-diagonal-traverse.cpp
class Solution {
public:
    vector<int> findDiagonalOrder(vector<vector<int>>& mat) {
        int m=mat.size(),n=mat[0].size();
        vector<int>res(m*n);
        bool up=true;
        int i=0;
        int rows=0,cols=0;
        while(rows<m && cols<n){
            
            if(up){
                while(rows>0 && cols<n-1){
                    res[i++]=mat[rows][cols];
                    rows--;
                    cols++;
                }
                res[i++]=mat[rows][cols];
                if(cols==n-1){
                    rows++;
                }
                else{
                    cols++;
                }
            }
            else{
                while(cols>0 && rows<m-1){
                    res[i++]=mat[rows][cols];
                    rows++;
                    cols--;
                }
                res[i++]=mat[rows][cols];
                if(rows==m-1){
                    cols++;
                }
                else{rows++;}
            }
            up=!up;
        }
        
        return res;
    }
};